### ISSUE TITLE
TLS Handshake record layer fragmentation not working

### ISSUE THREAD

**Issue by Irrialite** [Mbed-TLS/mbedtls#1840]
I have a server setup that sends handshake with `ssl->in_hslen` set to above 22700-ish, which is above the accepted limit of 16384. Is there any plan to support this (or a workaround without changing the server) and if not, I would appreciate some tips as to how to implement this as it is quite urgent for my project.

This is the code segment that fails:
`/* With TLS we don't handle fragmentation (for now) */
if( ssl->in_msglen < ssl->in_hslen )
{
MBEDTLS_SSL_DEBUG_MSG( 1, ( "TLS handshake fragmentation not supported" ) );
return( MBEDTLS_ERR_SSL_FEATURE_UNAVAILABLE );
}`

**Comment by RonEld** [comment]
@Irrialite As you can see from [this PR](https://github.com/ARMmbed/mbedtls/pull/1824) we are working an refactoring the messaging infrastructure. Message fragmentation being one of the issues addressed.
Unfortunately, as you can see, this is still work in progress, and we cannot commit on when it will be ready.

**Comment by RonEld** [comment]
You are welcome to try this PR and ad your comments though.

**Comment by RonEld** [comment]
Note that this PR does require integration with the rest of the library though.

**Comment by RonEld** [comment]
@Irrialite I have just looked again at your query.
Please note that TLS specification allows only up to 16 KB for the record data, so even once we support message fragmentation, sending 22700-ish is not allowed.
The application should only send 16 KB of data. If the data is larger, then perhaps the application should send it in several chunks, and the remote peer application should be able to reconstruct the whole data buffer.

**Comment by RonEld** [comment]
@Irrialite I would like to clarify, that my previous comment was about the Application data. I now realize you were asking about the Handshake message length. The standard doesn't limit the length of the handshake message( except that the length field is 24 bits, so the message cannot exceed 32MB).

In addition, the Fragmentation support which is being worked on, is about the DLTS fragmentation. Supporting TLS fragmentation is not at the moment planned, as usually it's not a problem.
Usually, the bottle neck of the handshake messages are the certificate messages. Is it possible that your server would send a shorter certificate chain? Have you considered using ECC based certificates?

**Comment by RonEld** [comment]
We believe this issue has been addressed, therefore closing.
If you believe this issue remains unclear or unanswered, please reopen, with an update

**Comment by RonEld** [comment]
Reopening, for the sake of tracking the TLS handshake fragmentation feature

**Comment by rojer** [comment]
@RonEld this is more serious when Max Fragment Length is used. if a device (using mbedTLS) asks for a small fragment, e.g. 1024 bytes, and the server complies, then handshake ends up fragmented:
```
handshake message: msglen = 1024, type = 11, hslen = 2675
```
which leads to the "fragmented handshake" failure.

**Comment by RonEld** [comment]
@rojer Thank you for your example.
You example is exactly using the max fragment length. (MFL)
MFLk is cucrrently supported for DTLS only.
I suggest you look at [this article](https://tls.mbed.org/kb/how-to/controlling_package_size) for mpore information.
In TLS, however, since the bottle neck is probably the certificate message, it is recommended you set your `MBEDTLS_SSL_IN_CONTENT_LEN` value to fit the certificate size. in your case, probably bigger than 2675.

**Comment by bl4ck5un** [comment]
@RonEld I'm running into the same problem while working on a **client**. So `mbedtls_ssl_conf_max_frag_len` is not supported for client either? I tried to set `MBEDTLS_SSL_IN_CONTENT_LEN` to `5120` without calling `mbedtls_ssl_conf_max_frag_len`. But the server still sends me messages longer than that.

**Comment by bl4ck5un** [comment]
Reading the code, I called `mbedtls_ssl_conf_max_frag_len(&conf, MBEDTLS_SSL_MAX_FRAG_LEN_NONE)` as well, thinking that's going to cause the client to negotiate with the server, which apparently didn't happen. Am I missing anything?

**Comment by oesh** [comment]
We (Facebook networking team) have started working on adding handshake fragmentation support to mbedTLS. We are going to upstream the patch to the master repository once it is ready.

**Comment by rojer** [comment]
that would be great! looking forward to it. would be super nice to also have it backported to 2.16

**Comment by bl4ck5un** [comment]
@oesh that will be wonderful!! I've been blocked by this for more than a year. How is your implementation going? Do you have a public dev branch that I might try?

**Comment by oesh** [comment]
@bl4ck5un hey, I was busy with something else, but I am back to it. I do have a public dev branch, and it will be terrific to see early reports from the users. I will post an update here once I have something that passes tests, probably in the course of the following two weeks.

**Comment by oesh** [comment]
> that would be great! looking forward to it. would be super nice to also have it backported to 2.16

@rojer I am starting with `2.16.6`, since this is what we need internally.

**Comment by oesh** [comment]
Sharing some of the design considerations. Please comment as appropriate.

# Memory management
MbedTLS expects the handshake message to be in `ssl->in_buf`. `ssl->in_buf` contains `MBEDTLS_SSL_BUFFER_LEN`, which is sufficient for a single TLS record. The fragmented handshake message will not necessarily fit in `MBEDTLS_SSL_BUF_LENGTH`. 

I see several design alternatives, from the simplest to the more complicated. All assume a new config constant, `MBEDTLS_SSL_HS_MSG_MAX_LEN`, which limits the size of a single handshake message.

## Simplest 
The simplest alternative is to increase the `MBEDTLS_SSL_BUFFER_LEN` to be sufficient for the individual handshake message. 

### Advantages

1. No changes to the TLS state machine. Easier to reason about.

### Disadvantages

1. The runtime memory footprint of a session will be permanently increased, by `3 * MBEDTLS_SSL_HS_MSG_MAX_LEN - MBEDTLS_SSL_MAX_CONTENT_LEN`, which is likely to be `3 * (64K - 16K) == 144K`. 

## Slight memory footprint optimization 
The second alternative is to increase only the size of `ssl->in_msg`, but to keep `ssl->out_msg` and `ssl->compress_buf` without change. 

### Advantages

1. No changes to the TLS state machine. Easier to reason about.
2. Compared to the first alternative, the runtime footprint penalty will be lower. 

### Disadvantages

1. The runtime memory footprint of a session will be permanently increased, by `MBEDTLS_SSL_HS_MSG_MAX_LEN - MBEDTLS_SSL_MAX_CONTENT_LEN`, which is likely to be `64K - 16K == 48K`. 
2. The code is using a single constant for the main input/output buffers, `MBEDTLS_SSL_BUFFER_LEN`. Increasing the size of just the`ssl->in_buf` will require keeping two different constants, e.g. `MBEDTLS_SSL_INGRESS_BUFFER_LEN` and `MBEDTLS_SSL_EGRESS_BUFFER_LEN`. 

## Agressive memory footprint optimization
Instead of passing `ssl->in_msg` to the TLS state machine, use a different pointer, e.g. `ssl->in_hs_msg`, which can point to either the `ssl->in_msg`, if the handshake message was not fragmented, or to a different buffer, e.g. `ssl->handshake->hs_msg` (similar to the DTLS handshake reassembly code).

### Advantages

1. The memory footprint will likely to be minimal.

### Disadvantages

1. This is by far the most complex approach, and it will be harder to debug and maintain.

## KISS (for now)

Considering the above alternatives, I am opting for the simplest alternative for the time being. My reasoning is:

1. The reassembly of the handshake messages can be disabled in the config parameters, hence the users who do not need the reassembly will suffer no runtime penalty.
2. The simplest alternative is by far the easiest to get right, and will result in a smaller diff.
3. It will be still possible to add memory optimizations later, in an incremental fashion.

**Comment by Vitaliy69** [comment]
Any progress?

Call on my IoT STM32 device:

`mbedtls_ssl_conf_max_frag_len(&conf, MBEDTLS_SSL_MAX_FRAG_LEN_512);`

and got the error:

```
Performing the SSL/TLS handshake.../build/mbedtls-mygkMM/mbedtls-2.16.4/library/ssl_tls.c:3698: 0x7ffd936d4c80: TLS handshake fragmentation not supported
/build/mbedtls-mygkMM/mbedtls-2.16.4/library/ssl_tls.c:4369: 0x7ffd936d4c80: mbedtls_ssl_handle_message_type() returned -28800 (-0x7080)
```

**Comment by oesh** [comment]
Update:

Support for fragmented TLS handshake messages is ready for a merge into the `development` branch. At the moment I am waiting for the final sign-off by the MbedTLS core devs. 

In the meantime, the implementation is available in https://github.com/oesh/mbedtls/tree/hs_fragmentation__design_doc 

If you want to try this out:

1. Ensure that `MBEDTLS_SSL_HS_DEFRAG_CLT` is enabled in `include/mbedtls/config.h`
2. If needed, adjust the maximal size of an individual handshake message via MBEDTLS_SSL_HS_DEFRAG_MAX_MSG_LENGTH. The default maximal size is 16384.
3. Enable defragmentation in the runtime by invoking `mbedtls_ssl_conf_hs_defrag_max_len( conf, len )` when setting up the TLS context. The `len` parameter is the maximal size of a handshake message, which should not exceed the `MBEDLTS_SSL_HS_DEFRAG_MAX_MSG_LENGTH` configuration parameter.

**Comment by oesh** [comment]
> > that would be great! looking forward to it. would be super nice to also have it backported to 2.16
> 
> @rojer I am starting with `2.16.6`, since this is what we need internally.

Update - I have asked the MbedTLS community whether there's any issue with backporting the feature to 2.16. The response was that since this is not a bug fix, it would be best to not risk the stability of the long-term-support versions:

https://lists.trustedfirmware.org/pipermail/mbed-tls/2020-October/000205.html

**Comment by rojer** [comment]
fwiw, Cesanta agreed to open source our mbedtls modifications, among them i have incoming handshake message defragmentation - https://github.com/mongoose-os/mbedtls/commit/097ccf2ff88f30e894cd14c80409498bbb02fa10

this is against 2.16, and to complete the work fragmentation of outgoing messages is required. i do plan to finish this work at some point, but no firm ETA.

**Comment by lance-proxy** [comment]
Hi all, any updates on this issue above? It seems like message fragmentation is becoming more and more common and going to be problem for IoT devices to not have support for this.

**Comment by rnhmjoj** [comment]
It's particularly annoying because the mbedtls seems to make some webserver return a fragmented handshake, even if well below the 16K limit. This is one such handshake (should be easily reproduced by visiting any website using netlify.com):

![image](https://user-images.githubusercontent.com/2817565/127368542-a79663fc-b8fd-4a27-84a0-7bbc4a6581d9.png)

Do you know what is going on? I tried comparing mbedtls and openssl client hellos but I nothing stands out:

![image](https://user-images.githubusercontent.com/2817565/127369592-adad3e46-d63f-43e0-95f8-3d1640ee40b8.png)

In particular, it's not asking a max_fragment_length, which has been mentioned above.

**Comment by hanno-arm** [comment]
@rnhmjoj Nothing stands out to me either - note that it's every implementation's choice to pick a maximum record plaintext size. 

We have fragmentation working in our experimental TLS 1.3 implementation and are in the process of upstreaming it. cc @daverodgman

**Comment by lance-proxy** [comment]
@hanno-arm would love to give it a try on my device, could you point me to the branch for the `experimental TLS 1.3 implementation` ?

**Comment by hanno-arm** [comment]
@lance-proxy The branch is currently still on @hannestschofenig's fork https://github.com/hannestschofenig/mbedtls/tree/tls13-prototype/, but we'll move it to the main repository soon. Please let me know if you have any issues - the quickest way is probably to open an issue on the fork.

**Comment by ashesman** [comment]
Has there been any further movement on this?  It seems to me rare that certificates are small enough to be sent without fragmentation, especially if smaller fragmentation sizes are used.  At the moment you can not take advantage of fragmentation for reduced memory foot print unless you can be sure that certificates will be small.  TLS is required to support fragmentation starting from handshake so seems like it would be a good thing to have.

**Comment by rojer** [comment]
I've put up https://github.com/Mbed-TLS/mbedtls/pull/8981 that fixes this. Feedback/test reports would be appreciated.

**Comment by zliucd** [comment]
This may often happen when TLS server sends certificates in different and sequential packets. We can use ```len``` field from TLS record header to identify missing bytes and parse them.  

Though I have solved this issue in my personal project(https://github.com/zliucd/TLSfeat), but why fragmentation happens is still unknown.

**Comment by rojer** [comment]
> why fragmentation happens is still unknown

wdym unknown? handshake fragmentation happens because it is being requested by the client.

**Comment by hauke** [comment]
Meta is using handshake fragmentation now. The TLS handshakes with https://www.facebook.com/ and https://www.instagram.com with mbedtls as client are failing for me with `TLS handshake fragmentation not supported`. Is there a workaround to prevent Meta and others from using handshake fragmentation?

**Comment by rojer** [comment]
@hauke they shouldn't be doing that if you don't ask for max fragment size. make sure you _don't_ set max frag size.

**Comment by hauke** [comment]
Meta is not violating rfc8446. A server is allowed to send fragmented handshake messages. 
>  Handshake messages MAY be coalesced into a single TLSPlaintext record or fragmented across several records

I do not know why they are doing it.

**Comment by zliucd** [comment]
Fragmentation saves back-forth packet data, but it's hard to tell the cause.

When you look into TLS code, there is literally no switch to turn on/off
fragmentation.  I guess it's mostly caused by TCP control.

Hauke Mehrtens ***@***.***> 于2024年12月23日周一 23:56写道：

**Comment by mpg** [comment]
I can reproduce the problem, as explained here: https://github.com/Mbed-TLS/mbedtls/issues/9861#issuecomment-2559316540 - increasing the debug level confirms that the client is _not_ sending a `max_fragment_length` extension so this is not it. The server is choosing for fragment the message for reasons of its own, which AFAICS is perfectly allowed by the standard.

I'm afraid I can't see any workaround.

I think we'll need to implement support for this. I'll note there has been two PRs about that: https://github.com/Mbed-TLS/mbedtls/pull/3817 and https://github.com/Mbed-TLS/mbedtls/pull/8981 but none of them are in a good shape right now.

We'll need to pick one to revive and complete (or create a new one). This is unlikely to be done by team members in the coming days considering most of us are or will soon be on holiday.

Note: we usually have a policy of not backporting new features (which, technically, this would be) to LTS branches, but I'm strongly inclined to make an exception here and backport this to  the 3.6 branch at least.

**Comment by mpg** [comment]
Things that might be worth trying out (I have no time at the moment):
- Does the problem persist if we force the TLS version to 1.2? If yes, that's a work-around (though not satisfactory of course).
- If not, does it go away with #8981 (which is only supposed to handle 1.2 for now)?

**Comment by rojer** [comment]
@mpg i can confirm TLS1.2 is a valid workaround:
 * `programs/ssl/ssl_client2 server_name=facebook.com server_port=443 request_page=/ debug_level=4 auth_mode=none crt_file=none` fails
 * `programs/ssl/ssl_client2 server_name=facebook.com server_port=443 request_page=/ debug_level=4 auth_mode=none crt_file=none max_version=tls12` works

**Comment by rojer** [comment]
@mpg https://github.com/Mbed-TLS/mbedtls/pull/9872 is the incoming messgae defragemntation part of https://github.com/Mbed-TLS/mbedtls/pull/8981, now updated to work with TLS 1.3. it's still lacking automated tests but can be tested manually and notable works with facebook servers now. it also applies cleanly to 3.6, so backporting should be easy, if desired.

**Comment by Neustradamus** [comment]
To follow this important ticket

**Comment by mpg** [comment]
Support for defragmenting TLS handshake messages has been added to the latest 3.6 release and also merged in development. Again, many thanks @rojer for providing the fix and working with us on this!

**Comment by rojer** [comment]
@mpg not that incoming message defragmentation has been merged, i suggest we discuss full support for smaller record sizes in https://github.com/Mbed-TLS/mbedtls/issues/10099

**Comment by pbizzarriINIM** [comment]
I think issue here is not fully fixed:

https://github.com/Mbed-TLS/mbedtls/issues/10787
